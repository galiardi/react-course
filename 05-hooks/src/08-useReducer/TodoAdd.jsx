import {useForm} from '../custom-hooks/useForm';

export const TodoAdd = ({onNewTodo}) => {

  const { todoInput, onInputChange, onResetForm} = useForm({todoInput: ""});
  
  const onFormSubmit = (e) => {
    e.preventDefault();
    if (todoInput.length < 2) return;
    onNewTodo({
      id: Date.now(),
      description: todoInput,
      done: false
    });
    onResetForm();   
  };
  return (
    <form onSubmit={onFormSubmit}>
      <input 
        type="text" 
        placeholder="debo..." 
        className="form-control"
        name="todoInput"
        value={todoInput}
        onChange={onInputChange}

      />
      <button 
        type="submit" 
        className="btn btn-outline-primary mt-1"
      >
        Agregar
      </button>
    </form>
  )
}
