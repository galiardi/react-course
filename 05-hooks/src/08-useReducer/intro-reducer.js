const initialState = [
  {
    id: 1,
    todo: 'learning react',
    done: false, 
  },
];

const todoReducer = (state = initialState, action = {}) => {

  switch (action.type) {
    case '[TODO] add todo':
      return [...state, action.payload]
  
    default:
      return state
  }

};

let todos = todoReducer();

const newTodo = {
  id: 2,
  todo: 'Hacer subdivision',
  done: false
}

const addTodoAction = {
  type: '[TODO] add todo',
  payload: newTodo
}

todos = todoReducer( todos, addTodoAction);

console.log(todos);
