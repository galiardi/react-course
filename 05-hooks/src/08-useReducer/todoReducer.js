export const todoReducer = ( state = [], action = {}) => {

  switch ( action.type ) {
    case '[TODO] Add Todo':
      return [...state, action.payload];
    
    case '[TODO] Delete Todo':
      return state.filter( todo => todo.id !== action.payload );
    
    case '[TODO] Toggle Todo':
      return state.map((todo) => {

        if (todo.id === action.payload) {
          return {
            ...todo,
            done: !todo.done
          }
        }
        
        return todo;
      })
      
    default:
      return state;
  }
}


// case '[TODO] Delete Todo':
//       const todoToDeleteIndex = state.findIndex((todo) => todo.id === action.payload);
//       const stateCopy = [...state];
//       stateCopy.splice(todoToDeleteIndex,1);
//       return stateCopy;

// case '[TODO] Toggle Todo':
//       const todo = state.find(todo => todo.id === action.payload);
//       const todoIndex = state.findIndex(todo => todo.id = action.payload);
//       const todoCopy = {...todo};
//       todoCopy.done = !todoCopy.done;
//       const newState = [...state];
//       newState.splice(todoIndex, 1, todoCopy);
//       return newState;