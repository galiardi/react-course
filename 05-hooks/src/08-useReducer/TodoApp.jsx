import { TodoAdd } from "./TodoAdd";
import { TodoList } from "./TodoList";
import { useTodo } from "./useTodo";

export const TodoApp = () => {

  const {todos, todosCount, pendingTodosCount, handleNewTodo, handleDeleteTodo, handleToggleTodo} = useTodo();
  
  return (

    <>
    <h1>TodoApp: {todosCount} todos, 
      <span>{ pendingTodosCount } pendientes</span>
    </h1>
    <hr />
    <div className="row">
      <div className="col-7">
        <TodoList 
          todos={todos}
          onDeleteTodo={ handleDeleteTodo }
          onToggleTodo={ handleToggleTodo }
        />
      </div>
      <div className="col-5">
        <h4>Agregar TODO</h4>
        <hr />
        <TodoAdd
          onNewTodo={handleNewTodo}
        />
      </div>
    </div>
    </>
    
  )
}
