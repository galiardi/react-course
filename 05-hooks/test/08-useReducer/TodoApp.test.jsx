import { fireEvent, render, screen } from "@testing-library/react";
import { TodoApp } from "../../src/08-useReducer/TodoApp";
import { useTodo } from '../../src/08-useReducer/useTodo';

jest.mock('../../src/08-useReducer/useTodo');

describe('Testing 08-useReducer/TodoApp', () => {

  useTodo.mockReturnValue({
    todos: [{
      id: 1,
      description: 'Learn react',
      done: false
    }], 
    todosCount: 1, 
    pendingTodosCount: 1,
    handleNewTodo: jest.fn(),
    handleDeleteTodo: jest.fn(),
    handleToggleTodo: jest.fn()
  });

  beforeEach(() => {
    jest.clearAllMocks();
  })
  
  test('Debe hacer match con el snapshot', () => {
    
    const { container } = render(<TodoApp />);
    expect(container).toMatchSnapshot();

  });

  
  test('Debe mostrar el componente correctamente', () => {

    render(<TodoApp />);
    screen.debug();
    
    expect(screen.getByText('Learn react')).toBeTruthy();
    expect(screen.getByRole('textbox')).toBeTruthy();
    
  });
  
 })