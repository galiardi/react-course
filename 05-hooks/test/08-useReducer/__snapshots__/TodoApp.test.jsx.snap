// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing 08-useReducer/TodoApp Debe hacer match con el snapshot 1`] = `
<div>
  <h1>
    TodoApp: 
    1
     todos,
    <span>
      1
       pendientes
    </span>
  </h1>
  <hr />
  <div
    class="row"
  >
    <div
      class="col-7"
    >
      <ul
        class="list-group"
      >
        <li
          class="list-group-item d-flex justify-content-between"
        >
          <span
            aria-label="span"
            class="align-self-center "
          >
            Learn react
          </span>
          <button
            class="btn btn-danger"
          >
            Borrar
          </button>
        </li>
      </ul>
    </div>
    <div
      class="col-5"
    >
      <h4>
        Agregar TODO
      </h4>
      <hr />
      <form>
        <input
          class="form-control"
          name="todoInput"
          placeholder="debo..."
          type="text"
          value=""
        />
        <button
          class="btn btn-outline-primary mt-1"
          type="submit"
        >
          Agregar
        </button>
      </form>
    </div>
  </div>
</div>
`;
